ws-server/src/server.rs

```diff
@@ -21,477 +21,201 @@
-	RegisterMethod {
-		/// Name of the method.
-		name: String,
-		/// Where to send requests.
-		handler: mpsc::Sender<(RawServerRequestId, jsonrpc::Params)>,
-	},
-
-	/// Send a response to a request that a client made.
-	AnswerRequest {
-		/// Request to answer.
-		request_id: RawServerRequestId,
-		/// Response to send back.
-		answer: Result<JsonValue, jsonrpc::Error>,
-	},
-
-	/// Registers a subscription. The server will then handle subscription requests of that
-	/// method.
-	RegisterSubscription {
-		/// Unique identifier decided by the front-end in order to identify this registered
-		/// subscription.
-		unique_id: usize,
-		/// Name of the method that registers the subscription.
-		subscribe_method: String,
-		/// Name of the method that unregisters the subscription.
-		unsubscribe_method: String,
-	},
-
-	/// Send out a notification to all the clients registered to a subscription.
-	SendOutNotif {
-		/// The value that was passed in [`FrontToBack::RegisterSubscription::unique_id`] earlier.
-		unique_id: usize,
-		/// Notification to send to the subscribed clients.
-		notification: JsonValue,
-	},
+		Ok(())
+	}
 }
 
 impl Server {
-	/// Initializes a new server.
-	pub async fn new(url: impl AsRef<str>) -> Result<Self, Box<dyn error::Error + Send + Sync>> {
-		let sockaddr: SocketAddr = url.as_ref().parse()?;
-		let transport_server = WsTransportServer::builder(sockaddr).build().await?;
-		let local_addr = *transport_server.local_addr();
-
-		// We use an unbounded channel because the only exchanged messages concern registering
-		// methods. The volume of messages is therefore very low and it doesn't make sense to have
-		// a backpressure mechanism.
-		// TODO: that's not true anymore ^
-		let (to_back, from_front) = mpsc::unbounded();
-
-		async_std::task::spawn(async move {
-			background_task(transport_server.into(), from_front).await;
-		});
-
-		Ok(Server {
-			to_back,
-			registered_methods: Arc::new(Mutex::new(Default::default())),
-			next_subscription_unique_id: Arc::new(atomic::AtomicUsize::new(0)),
-			local_addr,
-		})
+	pub fn register_method<F, R>(&mut self, method_name: &'static str, callback: F)
+	where
+		R: Serialize,
+		F: Fn(&str) -> Result<R, RpcError> + Send + Sync + 'static, // TODO: figure out correct lifetime here
+	{
+		self.methods.insert(
+			method_name,
+			Box::new(move |id, params, tx, _| {
+				let result = callback(params)?;
+
+				let json = serde_json::to_string(&JsonRpcResponse { jsonrpc: TwoPointZero, id, result })?;
+
+				tx.send(json).map_err(Into::into)
+			}),
+		);
 	}
 
-	/// Local socket address of the underlying transport server.
-	pub fn local_addr(&self) -> &SocketAddr {
-		&self.local_addr
-	}
+	pub fn register_subscription<T>(
+		&mut self,
+		subscribe_method_name: &'static str,
+		unsubscribe_method_name: &'static str,
+	) -> SubsciptionSink {
+		let subscribers = Arc::new(Mutex::new(FxHashMap::default()));
```

Comment:
is a `Mutex` required here because the different closures are `Send + Sync`? Otherwise, I don't understand why a Mutex is required.

Reply:
It needs a `Mutex` because you you should be able to share a sink across threads (including clones), plus the closures internally need access to the map (to sub or unsub).